src/Collection.hs

```diff
@@ -10,36 +10,79 @@
 
 
 module Collection(
-  Collection
-, choose
-, Collectable(..)
+    Collection
+  , choose
+  , Collectable(..)
+  , WeightedList
+  , weightedList
+  , survey
 ) where
 
-import Control.Monad.Trans.State.Strict(State)
-import System.Random(StdGen)
+import Control.Monad.Trans.State.Strict(State, state, runState)
+import Data.Tuple.Extra(second)
+import System.Random(StdGen, randomR, mkStdGen)
 
 import Random(pickItem)
 
 
 data Collection a = CollectionRaw a
                   | CollectionList [Collection a]
                   | CollectionState (State StdGen (Collection a))
+                  | CollectionWL (WeightedList (Collection a))
 
 
 choose :: Collection a -> State StdGen a
-choose (CollectionRaw a)   = return a
-choose (CollectionList aa) = pickItem aa >>= choose
-choose (CollectionState f) = f >>= choose
+choose (CollectionRaw a)                         = return a
+choose (CollectionList aa)                       = pickItem aa >>= choose
+choose (CollectionState s)                       = s >>= choose
+choose (CollectionWL (WeightedList items total)) = let
+    getItemAtWeight = getItemAtWeight' 0
```

Comment:
Simplify: this is a local helper already, so just put the 0 in the one function call and get rid of the sub-helper.